lib/rules/camelcase.js

```diff
@@ -5,6 +5,28 @@
 
 "use strict";
 
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const { findVariable } = require("eslint-utils");
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * Determines whether the given identifier node is a reference to a global variable.
+ * @param {ASTNode} node `Identifier` node to check.
+ * @param {Scope} scope Scope to which the node belongs.
+ * @returns {boolean} True if the identifier is a reference to a global variable.
+ */
+function isGlobalReference(node, scope) {
+    const variable = findVariable(scope, node);
+
+    return variable !== null && variable.scope.type === "global" && variable.defs.length === 0;
+}
```

Comment:
`eslint-utils.findVariable` searches for a variable by its name (it takes `node.name`) and doesn't check if the node is really a reference to the found variable.

This works fine for rules where the given identifier node is guaranteed to represent a variable reference.

Since we're checking all `Identifier` nodes in this rule, something like `isReferenceToGlobalVariable` implemented in #12987 might be more suitable for this particular rule.

You can also copy and modify test cases from that PR, these rules are similar (one disallows names with `_`, the other given list of names).

Reply:
I made the change, after rebasing onto master.
But I'm not sure about which tests to pull in.

Reply:
Also because I'm not very familiar with the test setup, in that module the options parameter for each test is different... I don't understand what I'm supposed to do.

Reply:
For example, the first test case from #12987 checks whether a disallowed name `Number` (    `options: ["Number"]` configures that rule to disallow `Number`) can be used to call a method on it:

```js
{
    code: "Number.parseInt()",
    options: ["Number"]
}
```

That test is `valid` only because `Number` is a global variable (that rule doesn't need `ignoreGlobals` option, it always allows the use of globals).

An equivalent test we could add now for the `camelcase` rule would be something like:

```js
{
    code: "snake_cased.foo()",
    options: [{ ignoreGlobals: true }],
    globals: { snake_cased: false } // eslint-disable-line camelcase
}
```

(this test will actually fail with the current iteration of this PR)

Let me know if you need further clarification or help with other tests.

Reply:
Okay, I added some of the tests that you mentioned. I gave a quick look to other tests in that rule, but I couldn't figure see any others that could apply. If there's any other let me know!

Reply:
Thanks! Can we also add these `valid` tests, I think they'll be useful for future maintenance:

```js
{
    code: "var foo = a_global_variable.bar",
    options: [{ ignoreGlobals: true }],
    globals: { a_global_variable: "readonly" } // eslint-disable-line camelcase
},
{
    code: "a_global_variable.foo = bar",
    options: [{ ignoreGlobals: true }],
    globals: { a_global_variable: "readonly" } // eslint-disable-line camelcase
},
{
    code: "( { foo: a_global_variable.bar } = baz )",
    options: [{ ignoreGlobals: true }],
    parserOptions: { ecmaVersion: 6 },
    globals: { a_global_variable: "readonly" } // eslint-disable-line camelcase
},
{
    code: "a_global_variable = foo",
    options: [{ ignoreGlobals: true }],
    globals: { a_global_variable: "writable" } // eslint-disable-line camelcase
},
{
    code: "a_global_variable = foo",
    options: [{ ignoreGlobals: true }],
    globals: { a_global_variable: "readonly" } // eslint-disable-line camelcase
},
{
    code: "({ a_global_variable } = foo)",
    options: [{ ignoreGlobals: true }],
    parserOptions: { ecmaVersion: 6 },
    globals: { a_global_variable: "writable" } // eslint-disable-line camelcase
},
{
    code: "({ snake_cased: a_global_variable } = foo)",
    options: [{ ignoreGlobals: true }],
    parserOptions: { ecmaVersion: 6 },
    globals: { a_global_variable: "writable" } // eslint-disable-line camelcase
},
{
    code: "({ snake_cased: a_global_variable = foo } = bar)",
    options: [{ ignoreGlobals: true }],
    parserOptions: { ecmaVersion: 6 },
    globals: { a_global_variable: "writable" } // eslint-disable-line camelcase
},
{
    code: "[a_global_variable] = bar",
    options: [{ ignoreGlobals: true }],
    parserOptions: { ecmaVersion: 6 },
    globals: { a_global_variable: "writable" } // eslint-disable-line camelcase
},
{
    code: "[a_global_variable = foo] = bar",
    options: [{ ignoreGlobals: true }],
    parserOptions: { ecmaVersion: 6 },
    globals: { a_global_variable: "writable" } // eslint-disable-line camelcase
},
{
    code: "foo[a_global_variable] = bar",
    options: [{ ignoreGlobals: true }],
    globals: { a_global_variable: "readonly" } // eslint-disable-line camelcase
},
{
    code: "var foo = { [a_global_variable]: bar }",
    options: [{ ignoreGlobals: true }],
    parserOptions: { ecmaVersion: 6 },
    globals: { a_global_variable: "readonly" } // eslint-disable-line camelcase
},
{
    code: "var { [a_global_variable]: foo } = bar",
    options: [{ ignoreGlobals: true }],
    parserOptions: { ecmaVersion: 6 },
    globals: { a_global_variable: "readonly" } // eslint-disable-line camelcase
},
```

Reply:
Done in https://github.com/eslint/eslint/pull/12782/commits/fd2cf42e69fe3daee898d61c910fd7963de4b8b0

Reply:
And the following `invalid` tests.

These are to test that the option has correct default value (`false`):

```js
{
    code: "var camelCased = snake_cased",
    globals: { snake_cased: "readonly" }, // eslint-disable-line camelcase
    errors: [
        {
            messageId: "notCamelCase",
            data: { name: "snake_cased" },
            type: "Identifier"
        }
    ]
},
{
    code: "var camelCased = snake_cased",
    options: [{}],
    globals: { snake_cased: "readonly" }, // eslint-disable-line camelcase
    errors: [
        {
            messageId: "notCamelCase",
            data: { name: "snake_cased" },
            type: "Identifier"
        }
    ]
}
```

These identifiers are not references:

```js
{
    code: "foo.a_global_variable = bar",
    options: [{ ignoreGlobals: true }],
    globals: { a_global_variable: "writable" }, // eslint-disable-line camelcase
    errors: [
        {
            messageId: "notCamelCase",
            data: { name: "a_global_variable" },
            type: "Identifier"
        }
    ]
},
{
    code: "var foo = { a_global_variable: bar }",
    options: [{ ignoreGlobals: true }],
    globals: { a_global_variable: "writable" }, // eslint-disable-line camelcase
    errors: [
        {
            messageId: "notCamelCase",
            data: { name: "a_global_variable" },
            type: "Identifier"
        }
    ]
},
{
    code: "var foo = { a_global_variable: a_global_variable }",
    options: [{ ignoreGlobals: true }],
    globals: { a_global_variable: "writable" }, // eslint-disable-line camelcase
    errors: [
        {
            messageId: "notCamelCase",
            data: { name: "a_global_variable" },
            type: "Identifier",
            column: 13
        }
    ]
},
{
    code: "var foo = { a_global_variable() {} }",
    options: [{ ignoreGlobals: true }],
    parserOptions: { ecmaVersion: 6 },
    globals: { a_global_variable: "writable" }, // eslint-disable-line camelcase
    errors: [
        {
            messageId: "notCamelCase",
            data: { name: "a_global_variable" },
            type: "Identifier"
        }
    ]
},
{
    code: "class Foo { a_global_variable() {} }",
    options: [{ ignoreGlobals: true }],
    parserOptions: { ecmaVersion: 6 },
    globals: { a_global_variable: "writable" }, // eslint-disable-line camelcase
    errors: [
        {
            messageId: "notCamelCase",
            data: { name: "a_global_variable" },
            type: "Identifier"
        }
    ]
},
{
    code: "a_global_variable: for (;;);",
    options: [{ ignoreGlobals: true }],
    globals: { a_global_variable: "writable" }, // eslint-disable-line camelcase
    errors: [
        {
            messageId: "notCamelCase",
            data: { name: "a_global_variable" },
            type: "Identifier"
        }
    ]
},
```

These are not references to a global variable, since it's shadowed:

```js
{
    code: "if (foo) { let a_global_variable; a_global_variable = bar; }",
    options: [{ ignoreGlobals: true }],
    parserOptions: { ecmaVersion: 6 },
    globals: { a_global_variable: "writable" }, // eslint-disable-line camelcase
    errors: [
        {
            messageId: "notCamelCase",
            data: { name: "a_global_variable" },
            type: "Identifier",
            column: 16
        },
        {
            messageId: "notCamelCase",
            data: { name: "a_global_variable" },
            type: "Identifier",
            column: 35
        }
    ]
},
{
    code: "function foo(a_global_variable) { foo = a_global_variable; }",
    options: [{ ignoreGlobals: true }],
    parserOptions: { ecmaVersion: 6 },
    globals: { a_global_variable: "writable" }, // eslint-disable-line camelcase
    errors: [
        {
            messageId: "notCamelCase",
            data: { name: "a_global_variable" },
            type: "Identifier",
            column: 14
        },
        {
            messageId: "notCamelCase",
            data: { name: "a_global_variable" },
            type: "Identifier",
            column: 41
        }
    ]
},
```

These are to show that the option does not apply to globals declared (or redeclared) in the given source code:

```js
{
    code: "var a_global_variable",
    options: [{ ignoreGlobals: true }],
    parserOptions: { ecmaVersion: 6 },
    errors: [
        {
            messageId: "notCamelCase",
            data: { name: "a_global_variable" },
            type: "Identifier"
        }
    ]
},
{
    code: "function a_global_variable () {}",
    options: [{ ignoreGlobals: true }],
    parserOptions: { ecmaVersion: 6 },
    errors: [
        {
            messageId: "notCamelCase",
            data: { name: "a_global_variable" },
            type: "Identifier"
        }
    ]
},
{
    code: "const a_global_variable = foo; bar = a_global_variable",
    options: [{ ignoreGlobals: true }],
    parserOptions: { ecmaVersion: 6 },
    globals: { a_global_variable: "writable" }, // eslint-disable-line camelcase
    errors: [
        {
            messageId: "notCamelCase",
            data: { name: "a_global_variable" },
            type: "Identifier",
            column: 7
        },
        {
            messageId: "notCamelCase",
            data: { name: "a_global_variable" },
            type: "Identifier",
            column: 38
        }
    ]
},
{
    code: "bar = a_global_variable; var a_global_variable;",
    options: [{ ignoreGlobals: true }],
    parserOptions: { ecmaVersion: 6 },
    globals: { a_global_variable: "writable" }, // eslint-disable-line camelcase
    errors: [
        {
            messageId: "notCamelCase",
            data: { name: "a_global_variable" },
            type: "Identifier",
            column: 7
        },
        {
            messageId: "notCamelCase",
            data: { name: "a_global_variable" },
            type: "Identifier",
            column: 30
        }
    ]
},
```